lib/spack/spack/hooks/sbang.py

```diff
@@ -186,42 +186,38 @@ def install_sbang():
     ``sbang`` here ensures that users can access the script and that
     ``sbang`` itself is in a short path.
     """
-    # copy in a new version of sbang if it differs from what's in spack
     sbang_path = sbang_install_path()
-    if os.path.exists(sbang_path) and filecmp.cmp(spack.paths.sbang_script, sbang_path):
-        return
 
-    # make $install_tree/bin
-    sbang_bin_dir = os.path.dirname(sbang_path)
-    fs.mkdirp(sbang_bin_dir)
+    all = spack.spec.Spec("all")
+    group_name = spack.package_prefs.get_package_group(all)
+    config_mode = spack.package_prefs.get_package_dir_permissions(all)
+    group_id = grp.getgrnam(group_name).gr_gid if group_name else None
 
-    # get permissions for bin dir from configuration files
-    group_name = spack.package_prefs.get_package_group(spack.spec.Spec("all"))
-    config_mode = spack.package_prefs.get_package_dir_permissions(spack.spec.Spec("all"))
+    # First setup the bin dir correctly.
+    sbang_bin_dir = os.path.dirname(sbang_path)
+    if not os.path.isdir(sbang_bin_dir):
+        fs.mkdirp(sbang_bin_dir)
 
-    if group_name:
-        os.chmod(sbang_bin_dir, config_mode)  # Use package directory permissions
+    # Set group and ownership like we do on package directories
+    if group_id:
+        os.chown(sbang_bin_dir, os.stat(sbang_bin_dir).st_uid, group_id)
+        os.chmod(sbang_bin_dir, config_mode)
     else:
         fs.set_install_permissions(sbang_bin_dir)
 
-    # set group on sbang_bin_dir if not already set (only if set in configuration)
-    if group_name and grp.getgrgid(os.stat(sbang_bin_dir).st_gid).gr_name != group_name:
-        os.chown(sbang_bin_dir, os.stat(sbang_bin_dir).st_uid, grp.getgrnam(group_name).gr_gid)
-
-    # copy over the fresh copy of `sbang`
-    sbang_tmp_path = os.path.join(
-        os.path.dirname(sbang_path),
-        ".%s.tmp" % os.path.basename(sbang_path),
-    )
-    shutil.copy(spack.paths.sbang_script, sbang_tmp_path)
+    # Then check if we need to install sbang itself.
+    try:
+        already_installed = filecmp.cmp(spack.paths.sbang_script, sbang_path)
+    except (IOError, OSError):
+        already_installed = False
 
-    # set permissions on `sbang` (including group if set in configuration)
-    os.chmod(sbang_tmp_path, config_mode)
-    if group_name:
-        os.chown(sbang_tmp_path, os.stat(sbang_tmp_path).st_uid, grp.getgrnam(group_name).gr_gid)
+    if not already_installed:
+        shutil.copy(spack.paths.sbang_script, sbang_path)
```

Comment:
This is still a race.  If two processes look at sbang and decide it's not there, one of these copies may end up appending or conflicting with the other.  We need *both* a unique tmpfile and the move into place.

I would tweak `llnl.util.filesystem.write_tmp_and_move` to be thread-safe (i.e. make it come up with a unique tmp name, which it doesn't currently do) and then use it.

Just use [`tempfile.NamedTemporaryFile`](https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile) with `prefix`, `suffix`, and `dir`:

```
>>> f = tempfile.NamedTemporaryFile(prefix=".foo.bar", suffix=".tmp", dir=".")
>>> f.name
'/Users/gamblin2/.foo.barhnlfjiqn.tmp'
```

Reply:
FWIW I could see this race coming up in CI for teams that clone spack and do a parallel install right away.

Reply:
Yeah, I think this is the best we can do. Confusingly, Linux has O_TMPFILE + linkat to accomplish pretty much what we need (invisible, auto-removed file, that can appear atomically after writing data into it), except that it's not able to rename, so, can't replace a file :(.